forch/mab_state_machine.py

```diff
@@ -0,0 +1,159 @@
+
+    def call_callbacks(self, call_list):
+        """Call list of callbacks"""
+        if call_list:
+            for callback in call_list:
+                callback()
+
+    def handle_timeout(self):
+        """Handle timeout"""
+        self.call_callbacks(self.on_timeout)
+
+    def exit_state(self):
+        """Call on_exit callbacks"""
+        self.call_callbacks(self.on_exit)
+
+
+class MacAuthBypassStateMachine():
+    """Class represents the MAB state machine that handles the MAB for a session"""
+
+    UNAUTH = "Unauthorized"
+    REQUEST = "Request sent"
+    IDLE = "Waiting RADIUS response"
+    ACCEPT = "Authorized"
+
+    LEARN = "Host learnt"
+    SENT_REQ = "Sent MAB request"
+    RECV_ACCPT = "Received RADIUS accept"
+    RECV_REJ = "Received RADIUS reject"
+    REQ_TIMEOUT = "RADIUS request timeout"
+    AUTH_TIMEOUT = "Auth session timed out"
+    EXPIRE = "Host expired"
+
+    def __init__(self, session):
+        self._retries = 0
+        self.session = session
+        self.states = {}
+        self._initialize_states()
+        self.current_state = self.states[self.UNAUTH]
+
+        self.transitions = [
+            {'trigger': self.LEARN, 'source': self.UNAUTH, 'dest': self.REQUEST},
+            {'trigger': self.SENT_REQ, 'source': self.REQUEST, 'dest': self.IDLE},
+            {'trigger': self.RECV_ACCPT, 'source': self.IDLE, 'dest': self.ACCEPT},
+            {'trigger': self.RECV_REJ, 'source': self.IDLE, 'dest': self.UNAUTH},
+            {'trigger': self.REQ_TIMEOUT, 'source': self.IDLE, 'dest': self.REQUEST,
+             'condition': self._retry_allowed, 'after': self._increment_retries},
+            {'trigger': self.REQ_TIMEOUT, 'source': self.IDLE,
+             'dest': self.UNAUTH, 'condition': self._too_many_retries},
+            {'trigger': self.AUTH_TIMEOUT, 'source': self.ACCEPT, 'dest': self.UNAUTH},
+            {'trigger': self.EXPIRE, 'source': '*', 'dest': self.UNAUTH}
+        ]
+
+    def _initialize_states(self):
+        self.states[self.UNAUTH] = State(name=self.UNAUTH, on_enter=self._handle_reject)
+        self.states[self.REQUEST] = State(name=self.REQUEST, on_enter=self._send_mab_request)
+        self.states[self.IDLE] = State(name=self.IDLE, timeout=self.session.response_timeout,
+                                       on_timeout=self._radius_timeout)
+        self.states[self.ACCEPT] = State(name=self.ACCEPT, on_enter=self._handle_accept,
+                                         timeout=self.session.session_timeout,
+                                         on_timeout=self._auth_session_timeout)
+
+    def _reset_state_machine(self):
+        self._retries = 0
+        self.current_state = self.states[self.UNAUTH]
+
+    def _radius_timeout(self):
+        self.process_trigger(self.REQ_TIMEOUT)
+
+    def _auth_session_timeout(self):
+        self.process_trigger(self.AUTH_TIMEOUT)
+
+    def _handle_accept(self):
+        self.session.session_result(accept=True)
```

Comment:
another indication that the other (session) class isn't helping -- here you just call straight into it. It would be clearner just to be able to say:  self.session_result(accept=True) without the intermediate